curvefs/src/metaserver/s3compact_wq_impl.cpp

```diff
@@ -0,0 +1,418 @@
+    const PartitionInfo& pinfo, const CopysetNode* copysetNode) {
+    uint64_t blockSize;
+    LOG_IF(FATAL, !conf_->GetUInt64Value("s3.blocksize", &blockSize));
+
+    // inode exist?
+    Inode inode;
+    MetaStatusCode ret = inodeStorage->Get(inodeKey, &inode);
+    if (ret != MetaStatusCode::OK) {
+        LOG(ERROR) << "GetInode fail, inodeKey = " << inodeKey.fsId << ","
+                   << inodeKey.inodeId
+                   << ", ret = " << MetaStatusCode_Name(ret);
+        return;
+    }
+
+    // deleted?
+    if (inode.nlink() == 0) {
+        return;
+    }
+
+    // s3chunkinfomap size 0?
+    if (inode.s3chunkinfomap().size() == 0) {
+        return;
+    }
+    uint64_t inodeId = inode.inodeid();
+    uint64_t inodeLen = inode.length();
+
+    // need compact?
+    const auto origMap(inode.s3chunkinfomap());
+    std::vector<uint64_t> needCompact(std::move(getNeedCompact(origMap)));
+    if (needCompact.empty()) {
+        return;
+    }
+
+    // let's compact
+    // 1. write new objs, each chunk one by one
+    std::vector<std::string> objsAdded;
+    std::vector<uint64_t> indexToDelete;
+    for (const auto& index : needCompact) {
+        // 1.1 build valid list
+        auto s3chunkVec =
+            inode.mutable_s3chunkinfomap()->at(index).mutable_s3chunks();
+        std::sort(s3chunkVec->begin(), s3chunkVec->end(),
+                  [](const S3ChunkInfo& a, const S3ChunkInfo& b) {
+                      return a.chunkid() < b.chunkid();
+                  });
+        const auto& s3chunkinfolist = origMap.at(index);
+
+        std::list<struct S3CompactWorkQueueImpl::Node> validList(
+            std::move(buildValidList(s3chunkinfolist, inodeLen)));
+        // 1.2 gen obj names and read s3
+        std::string fullChunk;
+        uint64_t newChunkid = 0;
+        uint64_t newCompaction = 0;
+        int ret = readFullChunk(validList, inodeId, blockSize, &fullChunk,
+                                &newChunkid, &newCompaction);
+        if (ret != 0) {
+            deleteObjs(objsAdded);
+            return;
+        }
+        // 1.3 then write objs with newChunkid and newCompaction
+        ret = writeFullChunk(fullChunk, inodeId, blockSize, newChunkid,
+                             newCompaction, &objsAdded);
+        if (ret != 0) {
+            deleteObjs(objsAdded);
+            return;
+        }
+        // 1.4 update inode locally
+        // record delete
+        indexToDelete.emplace_back(index);
+        // rm all
+        s3chunkVec->clear();
+        // add new
+        S3ChunkInfo toAdd;
+        toAdd.set_chunkid(newChunkid);
+        toAdd.set_compaction(newCompaction);
+        toAdd.set_offset(0);
+        toAdd.set_len(fullChunk.length());
+        toAdd.set_size(fullChunk.length());
+        toAdd.set_zero(false);
+        s3chunkVec->emplace_back(std::move(toAdd));
+    }
+
+    // 2. update inode
+    ret = updateInode(copysetNode, pinfo, inode);
+    if (ret != MetaStatusCode::OK) {
```

Comment:
if UpdateInode returns MetaStatusCode::REDIRECTED, you should skip all compact tasks about current copysetnode/partition.
or check whether current coyset node is leader in the begining of compact task, because copyset's leader can be transferred, even you have test it when S3CompactManager::Enqueue.

Reply:
each s3compact task  is inode based.
each Partition has a "class S3Compact", when "void S3CompactManager::Enqueue()" do a enqueue action to impl, it will check copysetNode->IsLeaderTerm() to keep i am the copyset node leader. if not, continue the loop but not enqueue.